File: src/agents/ContentGen.js
/**
 * ContentGen Agent — 文案生成
 * 重构版：按 docs/architecture/01-agent-design.md 规范
 */
import { globalEventBus } from '../core/EventBus.js';

export class ContentGenAgent {
    constructor({ llmProvider } = {}) {
        this.llmProvider = llmProvider;
        this.agentType   = 'CONTENT_GEN';
    }

    /**
     * @param {object} params
     * @param {object} params.product - { name, category, USP[] }
     * @param {object} params.target_persona - { age, interest[] }
     * @param {string[]} params.channels - ['tiktok', 'weibo']
     * @param {string}   params.tone - 'energetic' | 'professional' | 'warm'
     * @param {number}   params.ab_variants - 默认 3
     * @param {string}   params.campaign_id
     */
    async run(params) {
        const { product, target_persona, channels = [], tone = 'energetic',
                ab_variants = 3, campaign_id } = params;

        console.log(`[ContentGen] 📝 Generating ${ab_variants} variants for "${product?.name || 'product'}"`);

        const variants = await this._generateVariants({ product, target_persona, channels, tone, ab_variants });

        const output = {
            agent:    this.agentType,
            variants: variants,
            metadata: {
                language:       'zh-CN',
                word_count_avg: Math.round(variants.reduce((s, v) => s + v.body.length, 0) / variants.length),
                generated_at:   new Date().toISOString(),
            }
        };

        globalEventBus.publish('ContentGenerated', { bundle: output, campaign_id }, campaign_id);
        return output;
    }

    async _generateVariants({ product, target_persona, channels, tone, ab_variants }) {
        // 模拟 LLM 生成（生产环境替换为真实 API 调用）
        await this._simulateLatency(1500);

        const isTechnicalPromo = channels?.includes('zhihu') || channels?.includes('juejin');

        if (isTechnicalPromo) {
            return [{
                id:        'copy_vA',
                variant:   'A',
                title:     `深入浅出 ${product?.name || '开源项目'}：重构企业级数据分析工作流`,
                body:      `# ${product?.name || 'DataPulse'} 技术解析\n\n在现代数据驱动的商业环境中，我们需要更智能的工具。${product?.name} 正是为此而生。\n\n## 核心特性\n- ${product?.USP?.join('\n- ')}\n\n## 为什么选择 ${product?.name}？\n因为它不仅是一个工具，更是一个完整的生态。`,
                channel:   'zhihu',
                llm_model: 'claude-3-5-sonnet-mock',
            }];
        }

        const hooks = [
            `${product?.USP?.[0] || '全新功能'}，让你效率翻倍`,
            `${target_persona?.age || '年轻人'}都在用的${product?.name || '好工具'}`,
            `告别低效，${product?.name || '它'}来了`,
        ];

        return Array.from({ length: ab_variants }, (_, i) => ({
            id:        `copy_v${String.fromCharCode(65 + i)}`,  // copy_vA, copy_vB...
            variant:   String.fromCharCode(65 + i),              // 'A', 'B', 'C'
            hook:      hooks[i] || `版本 ${i + 1} 标题`,
            body:      `${product?.name || '产品'} 专为 ${target_persona?.age || '用户'} 设计，${product?.USP?.join('、') || '功能强大'}。立即体验！`,
            cta:       ['免费试用', '立即体验', '查看案例'][i] || '了解更多',
            tone,
            word_count: 45 + Math.floor(Math.random() * 30),
            llm_model: 'claude-3-5-sonnet-mock',
        }));
    }

    _simulateLatency(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

File: backend/app/agents/content_gen.py
"""
ContentGen Agent Node — LLM-powered A/B copy generation.

Input:  state.strategy, state.goal, state.constraints
Output: state.content  (bundle_id + variants[])
Events: ContentGenerated
"""
import json
import re
import uuid

import httpx
import structlog
from tenacity import retry, stop_after_attempt, wait_exponential

from app.config import settings
from app.core.event_bus import event_bus
from app.core.llm import llm_client
from .state import CampaignState

logger = structlog.get_logger(__name__)


async def _get_github_context(url: str) -> str:
    """
    Fetch comprehensive project context from GitHub:
    metadata, directory structure, recent commits, dependency files, README.
    Falls back gracefully — each section is independently optional.
    """
    match = re.match(r"https://github\.com/([^/]+)/([^/\s]+?)(?:\.git)?/?$", url)
    if not match:
        logger.warning("github_url_not_repo", url=url)
        return ""

    owner, repo = match.groups()
    api_base = f"https://api.github.com/repos/{owner}/{repo}"
    raw_base = f"https://raw.githubusercontent.com/{owner}/{repo}"
    api_headers = {"Accept": "application/vnd.github.v3+json"}
    parts: list[str] = []

    async with httpx.AsyncClient(timeout=30) as client:

        # ── 1. Repository metadata ────────────────────────────────────────
        try:
            r = await client.get(api_base, headers=api_headers)
            if r.status_code == 200:
                m = r.json()
                topics = ", ".join(m.get("topics") or []) or "N/A"
                parts.append(
                    f"[Meta] Stars:{m.get('stargazers_count', 0)} | "
                    f"Forks:{m.get('forks_count', 0)} | "
                    f"Language:{m.get('language', 'N/A')} | "
                    f"Topics:{topics} | "
                    f"Description:{m.get('description', '')}"
                )
                logger.info("github_meta_fetched", owner=owner, repo=repo)
        except Exception as exc:
            logger.warning("github_meta_failed", error=str(exc))

        # ── 2. Root directory structure ───────────────────────────────────
        try:
            r = await client.get(f"{api_base}/git/trees/HEAD", headers=api_headers)
            if r.status_code == 200:
                entries = [item["path"] for item in r.json().get("tree", [])]
                parts.append(f"[Structure] {' | '.join(entries[:50])}")
                logger.info("github_tree_fetched", entries=len(entries))
        except Exception as exc:
            logger.warning("github_tree_failed", error=str(exc))

        # ── 3. Recent commits (last 5) ────────────────────────────────────
        try:
            r = await client.get(f"{api_base}/commits?per_page=5", headers=api_headers)
            if r.status_code == 200:
                msgs = [c["commit"]["message"].split("\n")[0] for c in r.json()[:5]]
                parts.append(f"[Recent Commits] {' | '.join(msgs)}")
                logger.info("github_commits_fetched", count=len(msgs))
        except Exception as exc:
            logger.warning("github_commits_failed", error=str(exc))

        # ── 4. Dependency file (first match across branches) ──────────────
        dep_candidates = [
            "requirements.txt", "pyproject.toml",
            "package.json", "go.mod", "Cargo.toml",
        ]
        dep_found = False
        for fname in dep_candidates:
            if dep_found:
                break
            for branch in ("main", "master"):
                try:
                    r = await client.get(f"{raw_base}/{branch}/{fname}")
                    if r.status_code == 200:
                        parts.append(f"[{fname}]\n{r.text[:800]}")
                        logger.info("github_dep_fetched", file=fname, branch=branch)
                        dep_found = True
                        break
                except Exception:
                    continue

        # ── 5. README ─────────────────────────────────────────────────────
        readme = ""
        for branch in ("main", "master"):
            try:
                r = await client.get(f"{raw_base}/{branch}/README.md")
                if r.status_code == 200:
                    readme = r.text
                    logger.info("github_readme_fetched", branch=branch, length=len(readme))
                    break
            except Exception as exc:
                logger.warning("github_readme_fetch_failed", branch=branch, error=str(exc))

        if readme:
            parts.append(f"[README]\n{readme[:3000]}")

    return "\n\n".join(parts)


# NOTE: 2 attempts max — article generation is slow (~90s), 3 retries would hit ARQ 300s job limit.
@retry(stop=stop_after_attempt(2), wait=wait_exponential(multiplier=1, min=2, max=5))
async def _call_llm(prompt: str, is_article: bool = False) -> list[dict]:
    """Call LLM to generate copy variants. Retries 3x on transient errors."""
    system_prompt = (
        "You are a senior performance marketing copywriter and technical evangelist. "
        "Return a JSON array of copy variants."
    )

    if is_article:
        system_prompt += (
            " Generate exactly ONE article variant with fields: variant_label (set to 'A'), title, body, channel. "
            "The body must be plain text only — absolutely NO Markdown syntax: no ##, no **, no -, no `, no >. "
            "Structure the article like an academic paper: use numbered section headings like '1. 节名', '2. 节名', '2.1 小节名' on their own lines. "
            "Each section should have multiple paragraphs separated by a blank line. Write in depth — aim for 1500+ Chinese characters. "
            "IMPORTANT: The title MUST be 15 Chinese characters or fewer — count carefully, this is a hard limit."
        )
    else:
        system_prompt += (
            " Each variant must have: variant_label (A/B/C), title (MAX 15 chars), hook, body, cta, channel. "
        )

    system_prompt += "Output ONLY valid JSON, no markdown outside the JSON structure."

    raw = await llm_client.chat_completion(
        system=system_prompt,
        messages=[{"role": "user", "content": prompt}],
        max_tokens=8192 if is_article else 2048,
    )
    # Extract JSON array — find outermost [ ... ] to avoid being fooled
    # by code blocks (```python, ```yaml etc.) inside the article body.
    start = raw.find('[')
    end = raw.rfind(']')
    if start != -1 and end != -1:
        raw = raw[start:end + 1]

    try:
        return json.loads(raw)
    except json.JSONDecodeError as e:
        logger.error("content_gen_json_parse_error", error=str(e), raw_snippet=raw[:300])
        raise


async def content_gen_node(state: CampaignState) -> dict:
    """
    LangGraph node: generate A/B copy variants via Claude.
    """
    logger.info("content_gen_start", campaign_id=state["campaign_id"])

    strategy = state.get("strategy") or {}
    channels = strategy.get("channel_plan", [{"channel": "tiktok"}, {"channel": "meta"}])
    channel_names = [c["channel"] for c in channels]
    is_technical_promo = any(ch in ["zhihu", "juejin", "csdn"] for ch in channel_names)

    # Fetch comprehensive project context if a GitHub URL is present in goal
    repo_context = ""
    repo_url_match = re.search(r"https://github\.com/[^\s]+", state["goal"])
    if repo_url_match:
        repo_url = repo_url_match.group(0)
        repo_context = await _get_github_context(repo_url)

    prompt = (
        f"Product goal: {state['goal']}\n"
        f"Target channels: {', '.join(channel_names)}\n"
        f"KPI target: {state.get('kpi', {}).get('metric', 'awareness')} = {state.get('kpi', {}).get('target', 'high')}\n"
    )

    if repo_context:
        prompt += f"\nProject Context:\n{repo_context}\n"

    if is_technical_promo:
        prompt += (
            "\nGenerate ONE comprehensive technical article (variant_label: 'A'). "
            "Return a JSON array with exactly 1 variant.\n"
            "CRITICAL RULE — title字数: 标题必须≤15个汉字，例如'DataPulse架构深度解析'(12字)是合法的，"
            "'DataPulse v3.1深度解析：从多模型到知识图谱'(超过15字)是不合法的。"
            "生成前请数清楚字数，超过15字必须重新起名。"
        )
    else:
        prompt += "\nGenerate 3 A/B/C copy variants optimized for these channels."

    try:
        variants = await _call_llm(prompt, is_article=is_technical_promo)

        bundle_id = uuid.uuid4()
        bundle_data = {
            "bundle_id": str(bundle_id),
            "variants": variants,
            "llm_model": settings.gemini_model,
        }

        # ── Persistence Layer ─────────────────────────────────────────────
        from app.database import async_session_factory
        from app.models.content import ContentBundle, Copy

        async with async_session_factory() as db:
            # Ensure campaign exists (it might be 'demo' in some contexts,
            # we should skip persistence for demo or handle it)
            campaign_id = state.get("campaign_id")
            DEMO_UUID = uuid.UUID("00000000-0000-0000-0000-000000000001")
            try:
                camp_uuid = DEMO_UUID if campaign_id == "demo" else uuid.UUID(campaign_id)

                new_bundle = ContentBundle(
                    id=bundle_id,
                    campaign_id=camp_uuid,
                    llm_model=settings.gemini_model,
                    generation_params={"tone": state.get("tone", "professional")},
                )
                db.add(new_bundle)

                for var in variants:
                    new_copy = Copy(
                        bundle_id=bundle_id,
                        campaign_id=camp_uuid,
                        variant_label=var.get("variant_label", "A"),
                        hook=var.get("title") or var.get("hook") or "Untitled",
                        body=var.get("body"),
                        cta=var.get("cta"),
                        channel=var.get("channel"),
                        status="GENERATED"
                    )
                    db.add(new_copy)

                await db.commit()
                logger.info("content_gen_persisted", bundle_id=str(bundle_id))
            except (ValueError, TypeError):
                logger.warning("content_gen_persistence_skipped", campaign_id=campaign_id)

        await event_bus.publish(
            "ContentGenerated",
            {"bundle": bundle_data},
            state["campaign_id"],
        )

        logger.info("content_gen_done", variants=len(variants))
        return {
            "content": bundle_data,
            "status": "PRODUCTION",
            "completed_tasks": ["content_gen"],
        }

    except Exception as exc:
        logger.error("content_gen_error", error=str(exc))
        return {"errors": [{"node": "content_gen", "error": str(exc)}]}
